lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp

```diff
@@ -0,0 +1,242 @@
+                     U &&...args) {
+  Context.Diags.diagnose(loc,
+                         diag, std::forward<U>(args)...);
+}
+
+namespace {
+  class DiagnoseInfiniteRecursion : public SILFunctionTransform {
+  public:
+    DiagnoseInfiniteRecursion() {}
+
+  private:
+    static bool hasRecursiveCallInPath(SILBasicBlock &Block,
+                                       SILFunction *Target) {
+      // Process all instructions in the block to find applies that reference
+      // the parent function.  Also looks through vtables for statically
+      // dispatched (witness) methods.
+      auto *TargetModule = Target->getModule().getSwiftModule();
+      for (auto &I : Block) {
+        auto *AI = dyn_cast<ApplyInst>(&I);
+        if (AI && AI->getCalleeFunction() && AI->getCalleeFunction() == Target)
+          return true;
+
+        if (FullApplySite FAI = FullApplySite::isa(&I)) {
+          // Don't touch dynamic dispatch.
+          if (isa<ObjCMethodInst>(FAI.getCallee()))
+            continue;
+
+          auto &M = FAI.getModule();
+          if (auto *CMI = dyn_cast<ClassMethodInst>(FAI.getCallee())) {
+            auto ClassType = CMI->getOperand()->getType();
+
+            // FIXME: If we're not inside the module context of the method,
+            // we may have to deserialize vtables.  If the serialized tables
+            // are damaged, the pass will crash.
+            //
+            // Though, this has the added bonus of not looking into vtables
+            // outside the current module.  Because we're not doing IPA, let
+            // alone cross-module IPA, this is all well and good.
+            auto *BGC = ClassType.getNominalOrBoundGenericNominal();
+            if (BGC && BGC->getModuleContext() != TargetModule) {
+              continue;
+            }
+
+            auto *F = getTargetClassMethod(M, ClassType, CMI);
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+
+          if (auto *WMI = dyn_cast<WitnessMethodInst>(FAI.getCallee())) {
+            SILFunction *F;
+            SILWitnessTable *WT;
+
+            std::tie(F, WT) = M.lookUpFunctionInWitnessTable(
+                WMI->getConformance(), WMI->getMember());
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+        }
+      }
+      return false;
+    }
+
+    using ExitBlockSet = llvm::SmallPtrSetImpl<SILBasicBlock *>;
+    static bool hasInfinitelyRecursiveApply(SILFunction &Fn,
+                                            SILFunction *TargetFn,
+                                            const ExitBlockSet &ExitBlocks) {
+      enum RecursiveState {
+        FoundRecursion          = 0,
+        FoundPathOutOfFunction  = 1,
+      };
+
+      // Set the entry block into the state map.
+      llvm::DenseMap<SILBasicBlock *, RecursiveState> States;
+      States[Fn.getEntryBlock()] = FoundPathOutOfFunction;
```

Comment:
I don't understand why you need to store the state for blocks. If the traversal reaches an exit block, then it succeeds. But then to avoid revisiting blocks you either need to traverse without a worklist--RPO order--or keep a DenseSet of visited blocks.

Reply:
>  If the traversal reaches an exit block, then it succeeds.

That's not enough.  If this style of traversal reaches the exit block then I know I have a path from the entry node to that exit block that does not call itself.  Revisiting blocks isn't a problem because they're only re-enqueued into the worklist when they change state.  RPO gets me the "here's an exit block" part of the exit condition, but it doesn't get me a constructive proof of the "there exists a path from the entry block" part.

Reply:
>  But then to avoid revisiting blocks you either need to traverse without a worklist

Revisiting blocks isn't a problem because they're only re-enqueued into the worklist when they change state.

Reply:
That all makes sense. As I mentioned in other comments, things to consider are:
- A short description with better name/comments on RecursiveState.
- Revisiting the ExitBlocks at the end is misleading.
- *If* you end up running RPO on most large functions anyway, you might as well use it.